File: script/make/protocol/cs/DailyProtocol.cs
public static class DailyProtocol
{
    public static void Encode(System.Text.Encoding encoding, System.IO.BinaryWriter writer, System.UInt16 protocol, System.Collections.Generic.Dictionary<System.String, System.Object> data) 
    {
        switch (protocol) 
        {
            case 12301:
            {
                return;
            }
            case 12302:
            {
                return;
            }
            case 12303:
            {
                // 日常ID
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int32)(System.UInt32)data["dailyId"]));
                return;
            }
            case 12304:
            {
                // 阶段ID
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int32)(System.UInt32)data["stageId"]));
                return;
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }

    public static System.Collections.Generic.Dictionary<System.String, System.Object> Decode(System.Text.Encoding encoding, System.IO.BinaryReader reader, System.UInt16 protocol) 
    {
        switch (protocol) 
        {
            case 12301:
            {
                // 统计列表
                var listLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var list = new System.Collections.Generic.List<System.Object>(listLength);
                while (listLength-- > 0)
                {
                    // Count
                    // 统计类型
                    var type = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // 今日数量
                    var todayNumber = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // object
                    var count = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"type", type}, {"todayNumber", todayNumber}};
                    // add
                    list.Add(count);
                }
                return new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"list", list}};
            }
            case 12302:
            {
                // 日常列表
                var listLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var list = new System.Collections.Generic.List<System.Object>(listLength);
                while (listLength-- > 0)
                {
                    // Daily
                    // 日常ID
                    var dailyId = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // 是否领取奖励
                    var isAward = reader.ReadByte();
                    // object
                    var daily = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"dailyId", dailyId}, {"isAward", isAward}};
                    // add
                    list.Add(daily);
                }
                // DailyActive
                // 奖励阶段ID
                var stageId = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                // 活跃度
                var score = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                // object
                var dailyActive = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"stageId", stageId}, {"score", score}};
                return new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"list", list}, {"dailyActive", dailyActive}};
            }
            case 12303:
            {
                // 结果
                var resultLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var result = encoding.GetString(reader.ReadBytes(resultLength));
                return new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"result", result}};
            }
            case 12304:
            {
                // 结果
                var resultLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var result = encoding.GetString(reader.ReadBytes(resultLength));
                return new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"result", result}};
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }
}

File: script/make/protocol/erl/daily_protocol.erl
-module(daily_protocol).
-export([decode/2, encode/2]).
-include("count.hrl").
-include("daily.hrl").

-spec decode(Protocol :: non_neg_integer(), Binary :: binary()) -> {ok, [integer() | binary() | list()]} | {error, Protocol :: non_neg_integer(), Binary :: binary()}.
decode(12301, _Rest_ = <<_/binary>>) ->
    {ok, {}};

decode(12302, _Rest_ = <<_/binary>>) ->
    {ok, {}};

decode(12303, _Rest_ = <<_/binary>>) ->
    <<Data:32, _DataRest_/binary>> = _Rest_,
    {ok, Data};

decode(12304, _Rest_ = <<_/binary>>) ->
    <<Data:32, _DataRest_/binary>> = _Rest_,
    {ok, Data};

decode(Protocol, Binary) ->
    {error, Protocol, Binary}.


-spec encode(Protocol :: non_neg_integer(), Data :: atom() | tuple() | binary() | list()) -> {ok, binary()} | {error, Protocol :: non_neg_integer(), Data :: atom() | tuple() | binary() | list()}.
encode(12301, Data) ->
    Data12301 = <<(encode_data_12301(<<>>, 0, Data))/binary>>,
    {ok, <<(byte_size(Data12301)):16, 12301:16, Data12301/binary>>};

encode(12302, {#daily{daily_id = DailyDailyId, is_award = DailyIsAward}, #daily_active{stage_id = DailyActiveStageId, score = DailyActiveScore}}) ->
    Data12302 = <<DailyDailyId:32, DailyIsAward:8, DailyActiveStageId:32, DailyActiveScore:32>>,
    {ok, <<(byte_size(Data12302)):16, 12302:16, Data12302/binary>>};

encode(12303, Data) ->
    Data12303 = <<(protocol:text(Data))/binary>>,
    {ok, <<(byte_size(Data12303)):16, 12303:16, Data12303/binary>>};

encode(12304, Data) ->
    Data12304 = <<(protocol:text(Data))/binary>>,
    {ok, <<(byte_size(Data12304)):16, 12304:16, Data12304/binary>>};

encode(Protocol, Data) ->
    {error, Protocol, Data}.

encode_data_12301(Acc = <<_/binary>>, Length, []) ->
    <<Length:16, Acc/binary>>;
encode_data_12301(Acc = <<_/binary>>, Length, [#count{type = Type, today_number = TodayNumber} | Data]) ->
    encode_data_12301(<<Acc/binary, Type:32, TodayNumber:32>>, Length + 1, Data).

